### DISCUSSION TITLE
Files Service rejects DICOM files

### DISCUSSION THREAD

**Discussion by calairoh** [mia-platform/community#183] [Q&A] [answered]
Hi!

We're trying to upload to `files-service` a DICOM file, that has the `.dcm` extension. Unfortunately, the `files-service` returns a `415` error. In the logs of the service, we find this message:
```
[...] ERROR (7 on files-service-...): Invalid file provided, content type is not accepted by include list
    reqId: "..."
    includeList: []
    originalFileExtension: "dcm"
    fileExtensionFromStream: "dcm"
    mimeFromFileExtension: false
    mimeFromStream: "application/dicom"
```

We also tried to set the included extensions using the `FILE_TYPE_INCLUDE_LIST` environment variable, but it didn't work. Instead, we received an even stranger message, in which `.dcm` extension is not accepted even if it is explicitly included in the list:
```
[...] ERROR (7 on files-service-...): Invalid file provided, content type is not accepted by include list
    reqId: "..."
    includeList: [
      "dcm",
      "pdf",
      "ecgexam"
    ]
    originalFileExtension: "dcm"
    fileExtensionFromStream: "dcm"
    mimeFromFileExtension: false
    mimeFromStream: "application/dicom"
```

**Comment by fredmaggiowski** [comment]
My guess is that the files-service is not properly matching the `application/dicom` mime-type, as you can see both logs have the `mimeFromFileExtension` flag set to `false`

This may be due to a bug of the service if the `application/dicom` content-type is standard (as it appears to be)

**Reply by calairoh** [reply]
Yes, it seems to be standard: https://www.iana.org/assignments/media-types/application/dicom

**Reply by calairoh** [reply]
We temporarily solved the problem downgrading the `files-service` to version `2.1.0`. With this version, the `.dcm` extension is recognized as valid. Before, we were using the `2.6.4` version. It seems that something has been broken in some minor or patch release.

**Comment by Epado** [comment]
We encountered a similar problem when uploading a svg file, with the same version of file service, here is the log;

[1675171273996] ERROR (7 on files-service-846688c747-tcvkc): Invalid file provided, content type is not accepted by include list
    reqId: "2c881481de5e9bac74b9daa82fdd3abd, 2c881481de5e9bac74b9daa82fdd3abd"
    includeList: []
    originalFileExtension: "svg"
    fileExtensionFromStream: "xml"
    mimeFromFileExtension: "image/svg+xml"
    mimeFromStream: "application/xml"

**Reply by Epado** [reply]
also in our case FILE_TYPE_INCLUDE_LIST is not used, so the service should behave as it should include any file extension

**Reply by fredmaggiowski** [reply]
Hi, without the `FILE_TYPE_INCLUDE_LIST` the services allows for any file, however the mime must match. The returned error is misleading on this matter :)

The fact that it does not match means that it's either a bug in the files-service or the wrong mime is being supplied in the request

**Reply by Epado** [reply]
what do you mean the mime must match what would you expect it to be ? in my case or the dicom case ?

**Reply by fredmaggiowski** [reply]
in both cases, in your case the log states that the mime assumed by the file extension is `image/svg+xml` however the file is interpreted as a `application/xml`.

<details><summary>SVG mime mismatch from log</summary>

    mimeFromFileExtension: "image/svg+xml"
    mimeFromStream: "application/xml"

</details>

A similar issue is what affects the dicom type where apparently it was impossible to get the mime type from the file extension (since false is returned). A third party library is used for mime interpretation therefore we may have to tune it a little bit for these edgecases

<details><summary>DCM mime mismatch from log</summary>

    mimeFromFileExtension: false
    mimeFromStream: "application/dicom"

</details>

---

Imho there are two different things here that are happening:

 1. some custom code should be added in the files service to address these edgecases since these file types that do not get properly recognized (assuming the svg is getting uploaded with the correct mime type (`svg+xml`)
 2. the error in case of mismatching mime types should be improved to tell the user that the issue is with mismatching types rather than the fact that the `FILE_TYPE_INCLUDE_LIST` does not contain the desired file type


Do you have time to propose a fix on these two issues? otherwise @albertotessarotto let's track these issues in our internal backlog.

**Reply by giovanni-motterle-mia-care** [reply]
Hi @albertotessarotto @fredmaggiowski , did you track this issue in your backlog?
I don't understand what you mean by "uploaded with the correct mime type" since i don't control the mime type in any way.
Why do we care about mime type matching after all, if i want to upload a pdf that's called file.jpg why not?

**Reply by fredmaggiowski** [reply]
in regard of this

> Why do we care about mime type matching after all, if i want to upload a pdf that's called file.jpg why not?

The files service allows for file type upload protection, meaning that if you want your users to only be able to upload images you can configure the service to only accept such file types; however we cannot rely on the extension only for this check to be done because otherwise a malicious actor could upload a file with a different extension, therefore the service will take care of inspecting the mime type and match it with the declared extension and with the file header (if the file is binary); if these three don't match the file is rejected.


>if i want to upload a pdf that's called file.jpg why not?

Because, if you're downloading files on a browser, a mismatched extension could be abused to run arbitrary code in the web page and being vulnerable to XSS of CSRF, something you want to prevent for sure :)

As for the planned activity I'll let @albertotessarotto answer

**Reply by albertotessarotto** [reply]
Thank you for the ping. After Console v11 release I think we will be able to better identify a due date to assign for these activities on files service, and we will inform you about.

**Reply by giovanni-motterle-mia-care** [reply]
Hello, are there any updates on your planning to fix this issue?

**Reply by peppemanzi** [reply, marked as answer]
It was fixed in file-service v2.9.0, but this thread wasn't updated. I close it.
